search/query/conjunction.go

```diff
@@ -31,87 +31,126 @@ import (
 
 // ConjuctionQuery finds documents which match at least one of the given queries.
 type ConjuctionQuery struct {
-	Queries []search.Query
+	queries   []search.Query
+	negations []search.Query
 }
 
 // NewConjunctionQuery constructs a new query which matches documents which match all
 // of the given queries.
 func NewConjunctionQuery(queries []search.Query) search.Query {
 	qs := make([]search.Query, 0, len(queries))
+	ns := make([]search.Query, 0, len(queries))
 	for _, query := range queries {
-		// Merge conjunction queries into slice of top-level queries.
-		q, ok := query.(*ConjuctionQuery)
-		if ok {
-			qs = append(qs, q.Queries...)
+		switch query := query.(type) {
+		case *ConjuctionQuery:
+			// Merge conjunction queries into slice of top-level queries.
+			qs = append(qs, query.queries...)
 			continue
+		case *NegationQuery:
+			ns = append(ns, query.query)
+		default:
+			qs = append(qs, query)
 		}
+	}
 
-		qs = append(qs, query)
+	if len(qs) == 0 && len(ns) > 0 {
+		// We need to ensure we have at least one non-negation query from which we can calculate
+		// the set difference with the other negation queries.
+		qs = append(qs, queries[0])
```

Comment:
hm are you guaranteed this itself won't be a negation query?

Reply:
also, can you drop the matchall stuff now?

Reply:
It certainly could be a negation query but that's ok since we'll execute it like we do normally for negation queries, the matchall and then diff. That's why I didn't want to drop the matchall logic, otherwise you wouldn't be able to execute a negation query by itself.

Reply:
ah fair.